libraries/AP_IOMCU/iofirmware/iofirmware.cpp

```diff
@@ -48,83 +47,206 @@ const AP_HAL::HAL& hal = AP_HAL::get_HAL();
  a value of 2 means test with reboot
 */
 #define IOMCU_ENABLE_RESET_TEST 0
+#ifndef IOMCU_SEND_TX_FROM_RX
+#define IOMCU_SEND_TX_FROM_RX 1
+#endif
+// enable timing GPIO pings
+#ifdef IOMCU_LOOP_TIMING_DEBUG
+#undef TOGGLE_PIN_DEBUG
+#define TOGGLE_PIN_DEBUG(pin) do { palToggleLine(HAL_GPIO_LINE_GPIO ## pin); } while (0)
+#endif
 
 // pending events on the main thread
 enum ioevents {
-    IOEVENT_PWM=1,
+    IOEVENT_PWM = EVENT_MASK(1),
+    IOEVENT_TX_BEGIN = EVENT_MASK(2),
+    IOEVENT_TX_END = EVENT_MASK(3),
 };
 
-static void dma_rx_end_cb(UARTDriver *uart)
+// see https://github.com/MaJerle/stm32-usart-uart-dma-rx-tx for a discussion of how to run
+// separate tx and rx streams
+static void setup_rx_dma(hal_uart_driver* uart)
 {
-    osalSysLockFromISR();
-    uart->usart->CR3 &= ~(USART_CR3_DMAT | USART_CR3_DMAR);
-
-    (void)uart->usart->SR;
-    (void)uart->usart->DR;
-    (void)uart->usart->DR;
+    uart->usart->CR3 &= ~USART_CR3_DMAR;
     dmaStreamDisable(uart->dmarx);
-    dmaStreamDisable(uart->dmatx);
-
-    iomcu.process_io_packet();
-
     dmaStreamSetMemory0(uart->dmarx, &iomcu.rx_io_packet);
     dmaStreamSetTransactionSize(uart->dmarx, sizeof(iomcu.rx_io_packet));
+    dmaStreamSetPeripheral(uart->dmarx, &(uart->usart->DR));
     dmaStreamSetMode(uart->dmarx, uart->dmarxmode    | STM32_DMA_CR_DIR_P2M |
                      STM32_DMA_CR_MINC | STM32_DMA_CR_TCIE);
     dmaStreamEnable(uart->dmarx);
     uart->usart->CR3 |= USART_CR3_DMAR;
+}
 
+static void setup_tx_dma(hal_uart_driver* uart)
+{
+    uart->usart->CR3 &= ~USART_CR3_DMAT;
+    dmaStreamDisable(uart->dmatx);
     dmaStreamSetMemory0(uart->dmatx, &iomcu.tx_io_packet);
     dmaStreamSetTransactionSize(uart->dmatx, iomcu.tx_io_packet.get_size());
+    // starting the UART allocates the peripheral statically, so we need to reinstate it after swapping
+    dmaStreamSetPeripheral(uart->dmatx, &(uart->usart->DR));
     dmaStreamSetMode(uart->dmatx, uart->dmatxmode    | STM32_DMA_CR_DIR_M2P |
                      STM32_DMA_CR_MINC | STM32_DMA_CR_TCIE);
+    // enable transmission complete interrupt
+    uart->usart->SR = ~USART_SR_TC;
+    uart->usart->CR1 |= USART_CR1_TCIE;
+
     dmaStreamEnable(uart->dmatx);
+
     uart->usart->CR3 |= USART_CR3_DMAT;
-    osalSysUnlockFromISR();
 }
 
-static void idle_rx_handler(UARTDriver *uart)
+static void dma_rx_end_cb(hal_uart_driver *uart)
+{
+    chSysLockFromISR();
+    uart->usart->CR3 &= ~USART_CR3_DMAR;
```

Comment:
auto &usart = *uart->usart;
usart.SR = xxx;

Reply:
These are all over the current code and I am worried that if I try changing them all I will break something, so I am going to leave this.